File: src/EventEngine.ts
/** Atomic Redis storage for durable application events. @module */
import * as NodeCrypto from "@effect/platform-node/NodeCrypto";
import * as Context from "effect/Context";
import * as Crypto from "effect/Crypto";
import * as Data from "effect/Data";
import * as Effect from "effect/Effect";
import * as Layer from "effect/Layer";
import * as Option from "effect/Option";
import * as Schema from "effect/Schema";
import * as EventRecord from "./EventRecord.js";
import eventEngineScript from "./lua/eventEngine.js";
import * as NodeRedisPool from "./NodeRedisPool.js";
import {
  RedisConnectionRoles,
  type RedisConnectionRolesService,
  RedisPool,
  type RedisPoolService,
} from "./RedisPool.js";

/** Queue policy is persisted on first use. All clients must agree on it. */
export interface Queue {
  readonly name: string;
  readonly onCompletion?: "delete" | "archive";
  /** Default event lifetime in milliseconds; null or omitted means indefinite. */
  readonly ttlMs?: number | null;
  /** Lifetime after settlement; null or omitted archives indefinitely. */
  readonly archiveRetentionMs?: number | null;
}

/** Engine namespace and maximum records per maintenance category per call. */
export interface Config {
  readonly prefix?: string;
  readonly maintenanceBatchSize?: number;
}

const ErrorCodeSchema = Schema.Literals([
  "InvalidInput",
  "ConfigurationConflict",
  "CapacityExceeded",
  "SubscriptionMissing",
  "LeaseLost",
  "EventNotActive",
  "CorruptStorage",
  "IndeterminateWrite",
  "IdentityError",
]);

/** Storage/ownership failure. IndeterminateWrite means the operation may have committed. */
export class EventEngineError extends Data.TaggedError("EventEngineError")<{
  readonly code: typeof ErrorCodeSchema.Type;
  readonly message: string;
  readonly cause?: unknown;
  readonly operation?: string;
  readonly eventId?: string;
}> {}

/** Result of acknowledging a retained or already deleted event. */
export type Acknowledgement = "acknowledged" | "already-acknowledged" | "gone";
/** A bounded maintenance pass, with a signal for immediately due remaining work. */
export interface MaintenanceResult {
  readonly processed: number;
  readonly pending: boolean;
}
/** Encoded delivery for adapter authors. Prefer EventQueue for typed payloads. */
export interface EncodedDelivery {
  readonly event: EventRecord.EncodedEvent;
  readonly leaseToken: string;
}
/** Fenced identity used for mutation. */
export interface Attempt extends EventRecord.Subscription {
  readonly id: string;
  readonly token: string;
}

/** Atomic operations. Crypto is required only when generating new identities. */
export interface Service {
  readonly subscribe: (
    queue: Queue,
    name: string,
  ) => Effect.Effect<EventRecord.Subscription, EventEngineError, Crypto.Crypto>;
  readonly unsubscribe: (
    queue: Queue,
    subscription: EventRecord.Subscription,
  ) => Effect.Effect<boolean, EventEngineError>;
  readonly emit: (
    queue: Queue,
    payload: string,
    ttlMs?: number | null,
  ) => Effect.Effect<EventRecord.EncodedEvent, EventEngineError, Crypto.Crypto>;
  readonly get: (
    queue: Queue,
    id: string,
  ) => Effect.Effect<EventRecord.EncodedEvent | null, EventEngineError>;
  readonly take: (
    queue: Queue,
    subscription: EventRecord.Subscription,
    leaseMs?: number,
  ) => Effect.Effect<EncodedDelivery | null, EventEngineError, Crypto.Crypto>;
  readonly acknowledge: (
    queue: Queue,
    attempt: Attempt,
  ) => Effect.Effect<Acknowledgement, EventEngineError>;
  readonly renew: (
    queue: Queue,
    attempt: Attempt,
    leaseMs: number,
  ) => Effect.Effect<void, EventEngineError>;
  readonly release: (
    queue: Queue,
    attempt: Attempt,
    delayMs?: number,
  ) => Effect.Effect<void, EventEngineError>;
  readonly maintain: (
    queue: Queue,
  ) => Effect.Effect<MaintenanceResult, EventEngineError>;
  readonly listArchived: (
    queue: Queue,
    options?: { readonly offset?: number; readonly limit?: number },
  ) => Effect.Effect<ReadonlyArray<string>, EventEngineError>;
}

/** Provide with layerNoDeps for a custom Redis adapter, or layer for Node. */
export class EventEngine extends Context.Service<EventEngine, Service>()(
  "@effectmq/core/EventEngine",
) {}

const newIdentity = Effect.fnUntraced(function* (): Effect.fn.Return<
  string,
  EventEngineError,
  Crypto.Crypto
> {
  return yield* (yield* Crypto.Crypto).randomUUIDv4.pipe(
    Effect.mapError(
      (cause) =>
        new EventEngineError({
          code: "IdentityError",
          message: "Could not generate event identity",
          cause,
        }),
    ),
  );
});

const invalid = (message: string): EventEngineError =>
  new EventEngineError({ code: "InvalidInput", message });
const corrupt = (cause: unknown, operation?: string): EventEngineError =>
  new EventEngineError({
    code: "CorruptStorage",
    message: "Invalid event storage reply",
    cause,
    operation,
  });

/** Validate a nonempty queue/subscription name, bounded to 256 UTF-8 bytes. */
const validateName = (name: string): Effect.Effect<void, EventEngineError> =>
  typeof name === "string" &&
  name.length > 0 &&
  name.isWellFormed() &&
  Buffer.byteLength(name, "utf8") <= 256
    ? Effect.void
    : Effect.fail(invalid("Names must contain 1 to 256 UTF-8 bytes"));

// Bound sums below cjson's exact integer range, including the Redis timestamp.
const maxDurationMs = 3_153_600_000_000;
const validateDuration = (
  value: number | null,
  name: string,
  minimum = 0,
): Effect.Effect<void, EventEngineError> =>
  value === null ||
  (Number.isSafeInteger(value) && value >= minimum && value <= maxDurationMs)
    ? Effect.void
    : Effect.fail(
        invalid(
          `${name} must be an integer between ${minimum} and ${maxDurationMs}, or null`,
        ),
      );

const queuePolicy = Effect.fnUntraced(function* (
  queue: Queue,
): Effect.fn.Return<string, EventEngineError> {
  yield* validateName(queue.name);
  const onCompletion = queue.onCompletion ?? "delete";
  if (onCompletion !== "delete" && onCompletion !== "archive")
    return yield* invalid("Unknown completion policy");
  const ttlMs = queue.ttlMs ?? null;
  const archiveRetentionMs = queue.archiveRetentionMs ?? null;
  yield* validateDuration(ttlMs, "ttlMs");
  yield* validateDuration(archiveRetentionMs, "archiveRetentionMs");
  return JSON.stringify({ onCompletion, ttlMs, archiveRetentionMs });
});
const checkSubscription = Effect.fnUntraced(function* (
  queue: Queue,
  subscription: EventRecord.Subscription,
): Effect.fn.Return<void, EventEngineError> {
  if (subscription.queue !== queue.name)
    return yield* invalid("Subscription belongs to another queue");
  yield* validateName(subscription.name);
  yield* validateName(subscription.generation);
});

const ReplySchema = Schema.Union([
  Schema.Struct({ ok: Schema.Literal(true), value: Schema.Unknown }),
  Schema.Struct({
    ok: Schema.Literal(false),
    code: ErrorCodeSchema,
    message: Schema.String,
  }),
]);
const SubscriptionReply = Schema.Struct({
  name: Schema.String,
  generation: Schema.String,
});
const DeliveryReply = Schema.NullOr(
  Schema.Struct({
    event: EventRecord.EncodedEventSchema,
    leaseToken: Schema.String,
  }),
);
const AckReply = Schema.Literals([
  "acknowledged",
  "already-acknowledged",
  "gone",
]);
const MaintenanceReply = Schema.Struct({
  processed: Schema.Number,
  pending: Schema.Boolean,
});

/** Construct an engine with an existing Redis command service. */
export const makeWithRedis = Effect.fnUntraced(function* (
  redis: RedisPoolService,
  config: Config = {},
  roles?: RedisConnectionRolesService,
): Effect.fn.Return<Service, EventEngineError> {
  const prefix = config.prefix ?? "~effectmq:events:v1";
  const batch = config.maintenanceBatchSize ?? 100;
  if (typeof prefix !== "string" || prefix.length === 0 || /[{}]/.test(prefix))
    return yield* invalid("prefix must be nonempty and contain no braces");
  if (!Number.isSafeInteger(batch) || batch < 1 || batch > 1000)
    return yield* invalid("maintenanceBatchSize must be between 1 and 1000");

  const call = Effect.fnUntraced(function* <
    S extends Schema.Top & { readonly DecodingServices: never },
  >(
    queue: Queue,
    operation: string,
    input: object,
    schema: S,
  ): Effect.fn.Return<S["Type"], EventEngineError> {
    const policy = yield* queuePolicy(queue);
    const root = `${prefix}:{${Buffer.from(queue.name, "utf8").toString("base64url")}}`;
    const connection =
      roles === undefined
        ? redis
        : operation === "maintain"
          ? roles.maintenance
          : ["take", "acknowledge", "renew", "release"].includes(operation)
            ? roles.worker
            : roles.producer;
    const raw = yield* connection
      .evalScript<unknown>(
        eventEngineScript,
        { numberOfKeys: 1 },
        root,
        operation,
        policy,
        JSON.stringify(input),
        String(batch),
      )
      .pipe(
        Effect.mapError(
          (cause) =>
            new EventEngineError({
              code: "IndeterminateWrite",
              operation,
              eventId:
                "id" in input && typeof input.id === "string"
                  ? input.id
                  : undefined,
              message: `Event ${operation} may have committed; inspect durable state before retrying`,
              cause,
            }),
        ),
      );
    const invalidReply = (cause: unknown) => corrupt(cause, operation);
    const parsed = yield* Effect.try({
      try: () => {
        if (typeof raw !== "string")
          throw new TypeError("Expected a JSON reply");
        return JSON.parse(raw) as unknown;
      },
      catch: invalidReply,
    });
    const reply = yield* Schema.decodeUnknownEffect(ReplySchema)(parsed).pipe(
      Effect.mapError(invalidReply),
    );
    if (!reply.ok)
      return yield* new EventEngineError({
        code: reply.code,
        message: reply.message,
        operation,
      });
    return yield* Schema.decodeUnknownEffect(schema)(reply.value).pipe(
      Effect.mapError(invalidReply),
    );
  });
  const mutateAttempt = Effect.fnUntraced(function* <
    S extends Schema.Top & { readonly DecodingServices: never },
  >(
    queue: Queue,
    operation: string,
    attempt: Attempt,
    schema: S,
    extra: object = {},
  ): Effect.fn.Return<S["Type"], EventEngineError> {
    yield* checkSubscription(queue, attempt);
    yield* validateName(attempt.id);
    yield* validateName(attempt.token);
    return yield* call(queue, operation, { ...attempt, ...extra }, schema);
  });

  return EventEngine.of({
    subscribe: Effect.fnUntraced(function* (queue, name) {
      yield* validateName(name);
      const generation = yield* newIdentity();
      const identity = yield* call(
        queue,
        "subscribe",
        { name, generation },
        SubscriptionReply,
      );
      return { queue: queue.name, ...identity };
    }),
    unsubscribe: Effect.fnUntraced(function* (queue, subscription) {
      yield* checkSubscription(queue, subscription);
      return yield* call(queue, "unsubscribe", subscription, Schema.Boolean);
    }),
    emit: Effect.fnUntraced(function* (queue, payload, ttlMs) {
      const ttl = ttlMs === undefined ? (queue.ttlMs ?? null) : ttlMs;
      yield* validateDuration(ttl, "ttlMs");
      if (typeof payload !== "string" || Buffer.byteLength(payload) > 1_500_000)
        return yield* invalid(
          "Encoded payload must be a string no larger than 1.5 MB",
        );
      const id = yield* newIdentity();
      return yield* call(
        queue,
        "emit",
        { id, queue: queue.name, payload, ttlMs: ttl },
        EventRecord.EncodedEventSchema,
      );
    }),
    get: Effect.fnUntraced(function* (queue, id) {
      yield* validateName(id);
      return yield* call(
        queue,
        "get",
        { id },
        Schema.NullOr(EventRecord.EncodedEventSchema),
      );
    }),
    take: Effect.fnUntraced(function* (queue, subscription, leaseMs = 30_000) {
      yield* checkSubscription(queue, subscription);
      if (typeof leaseMs !== "number")
        return yield* invalid("leaseMs must be numeric");
      yield* validateDuration(leaseMs, "leaseMs", 1);
      const token = yield* newIdentity();
      return yield* call(
        queue,
        "take",
        { ...subscription, leaseMs, token },
        DeliveryReply,
      );
    }),
    acknowledge: (queue, attempt) =>
      mutateAttempt(queue, "acknowledge", attempt, AckReply),
    renew: Effect.fnUntraced(function* (queue, attempt, leaseMs) {
      if (typeof leaseMs !== "number")
        return yield* invalid("leaseMs must be numeric");
      yield* validateDuration(leaseMs, "leaseMs", 1);
      yield* mutateAttempt(queue, "renew", attempt, Schema.Boolean, {
        leaseMs,
      });
    }),
    release: Effect.fnUntraced(function* (queue, attempt, delayMs = 0) {
      if (typeof delayMs !== "number")
        return yield* invalid("delayMs must be numeric");
      yield* validateDuration(delayMs, "delayMs");
      yield* mutateAttempt(queue, "release", attempt, Schema.Boolean, {
        delayMs,
      });
    }),
    maintain: (queue) => call(queue, "maintain", {}, MaintenanceReply),
    listArchived: Effect.fnUntraced(function* (
      queue,
      { offset = 0, limit = 100 } = {},
    ) {
      if (
        !Number.isSafeInteger(offset) ||
        offset < 0 ||
        !Number.isSafeInteger(limit) ||
        limit < 1 ||
        limit > 1000
      )
        return yield* invalid(
          "Archive offset must be nonnegative and limit between 1 and 1000",
        );
      return yield* call(
        queue,
        "listArchived",
        { offset, limit },
        Schema.Array(Schema.String),
      );
    }),
  });
});

/** Construct an engine using a caller-provided RedisPool. */
export const make = Effect.fnUntraced(function* (
  config: Config = {},
): Effect.fn.Return<Service, EventEngineError, RedisPool> {
  return yield* makeWithRedis(
    yield* RedisPool,
    config,
    Option.getOrUndefined(yield* Effect.serviceOption(RedisConnectionRoles)),
  );
});
/** Dependency-free engine layer; supply RedisPool separately. */
export const layerNoDeps = (
  config: Config = {},
): Layer.Layer<EventEngine, EventEngineError, RedisPool> =>
  Layer.effect(EventEngine, make(config));
/** Standard Node composition, including Redis and Crypto services. */
export const layer = (
  config: {
    readonly engine?: Config;
    readonly redis?: NodeRedisPool.RedisConfig;
  } = {},
): Layer.Layer<
  | EventEngine
  | Crypto.Crypto
  | Layer.Success<ReturnType<typeof NodeRedisPool.layer>>,
  EventEngineError | Layer.Error<ReturnType<typeof NodeRedisPool.layer>>
> =>
  layerNoDeps(config.engine).pipe(
    Layer.provideMerge(
      Layer.merge(NodeRedisPool.layer(config.redis), NodeCrypto.layer),
    ),
  );

File: src/lua/eventEngine.lua
-- Durable application events. All keys belong to the queue's Redis hash slot.
local root = KEYS[1]
local op = ARGV[1]
local policyJson = ARGV[2]
local input = cjson.decode(ARGV[3])
local batch = tonumber(ARGV[4])
local policy = cjson.decode(policyJson)
local clock = redis.call('TIME')
local now = tonumber(clock[1]) * 1000 + math.floor(tonumber(clock[2]) / 1000)
local null = cjson.null
local subscriptions = root .. ':subscriptions'
local removed = root .. ':removed'
local retired = root .. ':retired'
local deadlines = root .. ':deadlines'
local archives = root .. ':archives'
local archiveDeadlines = root .. ':archive-deadlines'

local function ok(value) return cjson.encode({ ok = true, value = value }) end
local function fail(code, message) return cjson.encode({ ok = false, code = code, message = message }) end
local function recordKey(id) return root .. ':event:' .. id end
local function deliveryKey(generation) return root .. ':delivery:' .. generation end
local function save(record) redis.call('SET', recordKey(record.id), cjson.encode(record)) end
local function load(id)
  local raw = redis.call('GET', recordKey(id))
  if not raw then return nil end
  local decoded, record = pcall(cjson.decode, raw)
  if not decoded or type(record) ~= 'table' then error('EVENT_CORRUPT: Invalid event JSON') end
  if record.version ~= 1 or record.id ~= id or type(record.recipients) ~= 'table'
    or type(record.payload) ~= 'string' or type(record.createdAt) ~= 'number'
    or type(record.resolvedAt) ~= 'number' or type(record.queue) ~= 'string'
    or (record.settledAt ~= null and type(record.settledAt) ~= 'number')
    or (record.archiveUntil ~= null and type(record.archiveUntil) ~= 'number')
    or (record.status ~= 'active' and record.status ~= 'completed' and record.status ~= 'expired')
    or (record.expiresAt ~= null and type(record.expiresAt) ~= 'number') then
    error('EVENT_CORRUPT: Invalid event record')
  end
  for generation, recipient in pairs(record.recipients) do
    if type(recipient) ~= 'table' or recipient.generation ~= generation
      or type(recipient.name) ~= 'string'
      or (recipient.leaseToken ~= null and type(recipient.leaseToken) ~= 'string')
      or (recipient.leaseUntil ~= null and type(recipient.leaseUntil) ~= 'number')
      or (recipient.status ~= 'pending' and recipient.status ~= 'acknowledged' and recipient.status ~= 'waived') then
      error('EVENT_CORRUPT: Invalid event recipient')
    end
  end
  return record
end
local function purge(record)
  for generation, _ in pairs(record.recipients) do
    redis.call('ZREM', deliveryKey(generation), record.id)
  end
  redis.call('DEL', recordKey(record.id))
  redis.call('ZREM', deadlines, record.id)
  redis.call('ZREM', archives, record.id)
  redis.call('ZREM', archiveDeadlines, record.id)
end
local function settle(record, status, at)
  record.status = status
  record.settledAt = at
  for generation, recipient in pairs(record.recipients) do
    redis.call('ZREM', deliveryKey(generation), record.id)
    recipient.leaseUntil = null
  end
  redis.call('ZREM', deadlines, record.id)
  if policy.onCompletion == 'delete' then
    purge(record)
  else
    if policy.archiveRetentionMs ~= null then
      record.archiveUntil = at + policy.archiveRetentionMs
    end
    if record.archiveUntil ~= null and record.archiveUntil <= now then
      purge(record)
    else
      save(record)
      redis.call('ZADD', archives, at, record.id)
      if record.archiveUntil ~= null then
        redis.call('ZADD', archiveDeadlines, record.archiveUntil, record.id)
      end
    end
  end
end
local function normalize(record)
  if record.status ~= 'active' then
    if record.archiveUntil ~= null and record.archiveUntil <= now then purge(record) end
    return
  end
  local pending = 0
  local changed = false
  for generation, recipient in pairs(record.recipients) do
    if recipient.status == 'pending' then
      local removedAt = tonumber(redis.call('HGET', removed, generation))
      if removedAt and (record.expiresAt == null or removedAt < record.expiresAt) then
        recipient.status = 'waived'
        recipient.leaseToken = null
        recipient.leaseUntil = null
        record.resolvedAt = math.max(record.resolvedAt, removedAt)
        redis.call('ZREM', deliveryKey(generation), record.id)
        changed = true
      else
        pending = pending + 1
      end
    end
  end
  if pending == 0 then
    settle(record, 'completed', record.resolvedAt)
  elseif record.expiresAt ~= null and record.expiresAt <= now then
    settle(record, 'expired', record.expiresAt)
  elseif changed then
    save(record)
  end
end
local function current(id)
  local record = load(id)
  if record then
    normalize(record)
    if redis.call('EXISTS', recordKey(id)) == 0 then return nil end
  end
  return record
end
local function registered()
  return redis.call('HGET', subscriptions, input.name) == input.generation
end
local function maintain()
  local processed = 0
  -- Allocate a budget to each kind, so a removal backlog cannot starve expiry.
  for _, id in ipairs(redis.call('ZRANGEBYSCORE', deadlines, '-inf', now, 'LIMIT', 0, batch)) do
    local record = load(id)
    if record then normalize(record) else redis.call('ZREM', deadlines, id) end
    processed = processed + 1
  end
  for _, id in ipairs(redis.call('ZRANGEBYSCORE', archiveDeadlines, '-inf', now, 'LIMIT', 0, batch)) do
    local record = load(id)
    if record then purge(record) else redis.call('ZREM', archiveDeadlines, id); redis.call('ZREM', archives, id) end
    processed = processed + 1
  end
  local remaining = batch
  for _, generation in ipairs(redis.call('ZRANGE', retired, 0, batch - 1)) do
    if remaining <= 0 then break end
    local ids = redis.call('ZRANGE', deliveryKey(generation), 0, remaining - 1)
    for _, id in ipairs(ids) do
      local record = load(id)
      if record then normalize(record) end
      -- A terminal event or a waiver no longer needs this generation's index.
      redis.call('ZREM', deliveryKey(generation), id)
      remaining = remaining - 1
      processed = processed + 1
    end
    if #ids == 0 then remaining = remaining - 1 end
    if redis.call('ZCARD', deliveryKey(generation)) == 0 then
      redis.call('ZREM', retired, generation)
      redis.call('HDEL', removed, generation)
    end
  end
  return { processed = processed, pending = redis.call('ZCARD', retired) > 0
    or redis.call('ZCOUNT', deadlines, '-inf', now) > 0
    or redis.call('ZCOUNT', archiveDeadlines, '-inf', now) > 0 }
end

local function dispatch()
  local existingPolicy = redis.call('GET', root .. ':config')
  if existingPolicy and existingPolicy ~= policyJson then
    return fail('ConfigurationConflict', 'Queue already exists with a different policy')
  end
  if not existingPolicy then redis.call('SET', root .. ':config', policyJson) end

  if op == 'subscribe' then
    local generation = redis.call('HGET', subscriptions, input.name)
    if not generation then
      if redis.call('HLEN', subscriptions) >= 1000 then return fail('CapacityExceeded', 'Queue has 1000 active subscriptions') end
      generation = input.generation
      redis.call('HSET', subscriptions, input.name, generation)
    end
    return ok({ name = input.name, generation = generation })
  elseif op == 'unsubscribe' then
    if not registered() then return ok(false) end
    redis.call('HDEL', subscriptions, input.name)
    redis.call('HSET', removed, input.generation, now)
    redis.call('ZADD', retired, now, input.generation)
    maintain()
    return ok(true)
  elseif op == 'emit' then
    if redis.call('EXISTS', recordKey(input.id)) == 1 then return fail('InvalidInput', 'Event id already exists') end
    local record = { version = 1, id = input.id, queue = input.queue, payload = input.payload,
      createdAt = now, resolvedAt = now, expiresAt = null, settledAt = null, archiveUntil = null,
      status = 'active', recipients = {} }
    if input.ttlMs ~= null then record.expiresAt = now + input.ttlMs end
    local members = redis.call('HGETALL', subscriptions)
    for i = 1, #members, 2 do
      local name, generation = members[i], members[i + 1]
      record.recipients[generation] = { name = name, generation = generation, status = 'pending', leaseToken = null, leaseUntil = null }
      redis.call('ZADD', deliveryKey(generation), now, input.id)
    end
    save(record)
    if record.expiresAt ~= null then redis.call('ZADD', deadlines, record.expiresAt, input.id) end
    normalize(record)
    return ok(record)
  elseif op == 'get' then
    return ok(current(input.id) or null)
  elseif op == 'take' then
    if not registered() then return fail('SubscriptionMissing', 'Subscription generation is no longer registered') end
    for _, id in ipairs(redis.call('ZRANGEBYSCORE', deliveryKey(input.generation), '-inf', now, 'LIMIT', 0, batch)) do
      local record = current(id)
      if record and record.status == 'active' then
        local recipient = record.recipients[input.generation]
        if recipient and recipient.status == 'pending' then
          recipient.leaseToken = input.token
          recipient.leaseUntil = now + input.leaseMs
          if record.expiresAt ~= null then recipient.leaseUntil = math.min(recipient.leaseUntil, record.expiresAt) end
          redis.call('ZADD', deliveryKey(input.generation), recipient.leaseUntil, id)
          save(record)
          return ok({ event = record, leaseToken = input.token })
        end
      end
      redis.call('ZREM', deliveryKey(input.generation), id)
    end
    return ok(null)
  elseif op == 'acknowledge' or op == 'renew' or op == 'release' then
    local record = current(input.id)
    if not record then
      if op == 'acknowledge' then return ok('gone') end
      return fail('LeaseLost', 'Event is no longer retained')
    end
    local recipient = record.recipients[input.generation]
    if not recipient or recipient.name ~= input.name or recipient.leaseToken ~= input.token then
      return fail('LeaseLost', 'Delivery token is not current')
    end
    if op == 'acknowledge' and recipient.status == 'acknowledged' then return ok('already-acknowledged') end
    if record.status ~= 'active' then return fail('EventNotActive', 'Event has settled') end
    if recipient.status ~= 'pending' or recipient.leaseUntil == null or recipient.leaseUntil <= now or not registered() then
      return fail('LeaseLost', 'Delivery lease has expired or subscription was removed')
    end
    if op == 'acknowledge' then
      recipient.status = 'acknowledged'
      recipient.leaseUntil = null
      record.resolvedAt = now
      redis.call('ZREM', deliveryKey(input.generation), input.id)
      save(record)
      normalize(record)
      return ok('acknowledged')
    elseif op == 'renew' then
      recipient.leaseUntil = now + input.leaseMs
      if record.expiresAt ~= null then recipient.leaseUntil = math.min(recipient.leaseUntil, record.expiresAt) end
      redis.call('ZADD', deliveryKey(input.generation), recipient.leaseUntil, input.id)
    else
      recipient.leaseToken = null
      recipient.leaseUntil = null
      redis.call('ZADD', deliveryKey(input.generation), now + input.delayMs, input.id)
    end
    save(record)
    return ok(true)
  elseif op == 'maintain' then
    return ok(maintain())
  elseif op == 'listArchived' then
    maintain()
    local minimum = '-inf'
    if policy.archiveRetentionMs ~= null then minimum = '(' .. tostring(now - policy.archiveRetentionMs) end
    local ids = redis.call('ZRANGEBYSCORE', archives, minimum, '+inf', 'LIMIT', input.offset, input.limit)
    -- cjson encodes empty Lua tables as objects; preserve the array reply explicitly.
    if #ids == 0 then return '{"ok":true,"value":[]}' end
    return ok(ids)
  end
  return fail('InvalidInput', 'Unknown event operation')

end
local success, result = pcall(dispatch)
if success then return result end
if string.find(tostring(result), 'EVENT_CORRUPT:', 1, true) then
  return fail('CorruptStorage', tostring(result))
end
-- Preserve unexpected script failures as errors; writes before them may have committed.
return redis.error_reply(tostring(result))
